torchdata/datapipes/iter/util/webdataset.py

```diff
@@ -102,3 +113,488 @@ def __iter__(self) -> Iterator[Dict]:
 
     def __len__(self) -> int:
         return len(self.source_datapipe)
+
+
+def shardexpand(s):
+    expansion = r"[{][0-9]+[.][.][0-9]+[}]"
+    m = re.search(expansion, s)
+    if not m:
+        return [s]
+    prefix = s[: m.start()]
+    rest = shardexpand(s[m.end() :])
+    rng = s[m.start() + 1 : m.end() - 1]
+    lohi = rng.split("..")
+    if len(lohi[0]) != len(lohi[1]):
+        raise ValueError("Shard specifications must have " + f"same number of digits for low and high values in {s}.")
+    lo, hi = [int(x) for x in lohi]
+    if lo >= hi:
+        raise ValueError(f"Bad range in in shard spec {s}.")
+    result = []
+    for i in range(lo, hi + 1):
+        for r in rest:
+            expanded = f"{prefix}{i:0>{len(lohi[1])}}{r}"
+            result.append(expanded)
+    return result
+
+
+@functional_datapipe("shardexpand")
+class ShardExpanderIterDataPipe(IterDataPipe[Dict]):
```

Comment:
This seems to be usefull. If possible please add non-zero followed option too:
ex: 
{01-14} generates 010,011,012,013,014
{10-14} generates 10,11,12,13,14
however
{10-014} should error as it is unclear what to do